File: clickeye-api/app/models/central_contract.py
"""중앙 계약 관리 모델.

CentralContract: 중앙에서 관리하는 실행 계약 정의
CustomerContractOverride: 고객 프로젝트별 허용 필드 오버라이드
ContractAuditLog: 계약 변경 감사 로그
"""

import uuid
from datetime import UTC, datetime

from sqlalchemy import JSON, Boolean, Column, DateTime, ForeignKey, String, Text, Uuid

from app.database import Base


class CentralContract(Base):
    """중앙 실행 계약 정의."""

    __tablename__ = "central_contracts"

    id = Column(Uuid, primary_key=True, default=uuid.uuid4)
    slug = Column(String(200), nullable=False, unique=True, index=True)
    contract_type = Column(String(50), nullable=False)
    source = Column(String(200), nullable=False)
    version = Column(String(50), nullable=False, default="1.0.0")
    content = Column(JSON, nullable=False, default=dict)
    description = Column(Text, nullable=True)
    is_locked = Column(Boolean, nullable=False, default=True)
    allowed_overrides = Column(JSON, nullable=False, default=list)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(UTC))


class CustomerContractOverride(Base):
    """고객 프로젝트별 계약 오버라이드."""

    __tablename__ = "customer_contract_overrides"

    id = Column(Uuid, primary_key=True, default=uuid.uuid4)
    project_id = Column(
        Uuid,
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    central_contract_id = Column(
        Uuid,
        ForeignKey("central_contracts.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    override_content = Column(JSON, nullable=False, default=dict)
    approved_by = Column(
        Uuid,
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    is_active = Column(Boolean, nullable=False, default=True)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(UTC))


class ContractAuditLog(Base):
    """계약 변경 감사 로그."""

    __tablename__ = "contract_audit_logs"

    id = Column(Uuid, primary_key=True, default=uuid.uuid4)
    contract_id = Column(
        Uuid,
        ForeignKey("central_contracts.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    override_id = Column(
        Uuid,
        ForeignKey("customer_contract_overrides.id", ondelete="SET NULL"),
        nullable=True,
    )
    actor_id = Column(
        Uuid,
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=False,
    )
    change_type = Column(String(50), nullable=False)
    diff_snapshot = Column(JSON, nullable=False, default=dict)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(UTC))

File: clickeye-contracts/protocol/central_contract.ts
/**
 * 24SevenClaw 중앙 실행 계약 관리 타입 정의
 * 중앙 레포(private)에서 settings, skills, agents, pipelines 계약을 관리하고
 * 고객 프로젝트에 배포. 허용된 필드만 오버라이드 가능.
 */

// === 계약 소스 & 타입 ===

export type ContractSource = 'central' | 'custom';

export type ContractType = 'settings' | 'skill' | 'agent' | 'pipeline';

// === 중앙 계약 ===

export interface CentralContract {
  id: string;
  slug: string;
  contract_type: ContractType;
  source: ContractSource;
  version: string;
  content: Record<string, unknown>;
  is_locked: boolean;
  allowed_overrides: string[];
  created_at?: string;
  updated_at?: string;
}

// === 고객 오버라이드 ===

export interface CustomerContractOverride {
  id: string;
  project_id: string;
  central_contract_id: string;
  override_content: Record<string, unknown>;
  approved_by: string | null;
  is_active: boolean;
  created_at?: string;
  updated_at?: string;
}

// === 계약 감사 로그 ===

export type ContractChangeType = 'create' | 'update' | 'delete' | 'apply' | 'override' | 'sync';

export interface ContractAuditEntry {
  id: string;
  contract_id: string | null;
  override_id: string | null;
  actor_id: string;
  change_type: ContractChangeType;
  diff_snapshot: Record<string, unknown> | null;
  created_at: string;
}

// === 에이전트 동기화 페이로드 ===

export interface ContractSyncPayload {
  project_id: string;
  contracts: ContractSyncItem[];
}

export interface ContractSyncItem {
  slug: string;
  contract_type: ContractType;
  version: string;
  content: Record<string, unknown>;
  overrides: Record<string, unknown>;
}
